File: quix/quix.py
import sqlite3
from random import randrange, choices


class QuixSql():
    def __init__(self, db_path="./dq.db"):
        self.db = sqlite3.connect(db_path).cursor()
        self._fragment = self.random_key()
        self._current_key = self._fragment
        self._values = None
        self._chapter = None
        self._final_fragment = False

    
    def generate_fragment(self):
        while (not self._final_fragment) and (len(self.values()) > 0):
            next_word = self.values().pop(randrange(len(self.values())))[0]
            possible_next_fragment = " ".join([self.fragment(), next_word])
            if self.fragment_chapter(possible_next_fragment):
                self._fragment = possible_next_fragment
                self._current_key = None
                self._values = None
                self.generate_fragment()
            else:
                pass
        self._final_fragment = True

        return {"fragment": self.fragment(), "chapter": self.fragment_chapter(self.fragment())[0]}

    def fragment_chapter(self, fragment):
        if self._chapter:
            return self.known_fragment_chapter(fragment)
        return self.not_yet_known_fragment_chapter(fragment)
    
    def not_yet_known_fragment_chapter(self, fragment):
        chapters =  self.db.execute("SELECT chapter from chapters where text like ?", (f"%{fragment}%",)).fetchall()
        if len(chapters) == 1:
            self._chapter = chapters[0][0] 
        return chapters[0] if chapters else None

    def known_fragment_chapter(self, fragment):
        chapter =  self.db.execute("SELECT chapter from chapters where text like ? and chapter=?", (f"%{fragment}%", self._chapter)).fetchone()
        return chapter

    def values(self):
        if self._values is None:
            self._values = self.values_from(self.current_key())
        return self._values

    def current_key(self):
        if self._current_key is None:
            new_key_elements = self.fragment().split(' ')[-2:]
            self._current_key = " ".join(new_key_elements)
        return self._current_key

    def random_key(self, db=None):
        _db = db if db else self.db
        key =  _db.execute("SELECT key from ENGLISH ORDER BY RANDOM() LIMIT 1").fetchone()[0]
        return key

    def fragment(self):
        return self._fragment

    def values_from(self, key):
        values = self.db.execute("SELECT value, count from ENGLISH where key=?", (key,)).fetchall()
        return values

    def random_value_from(self, key):
        words_and_counts = self.values_from(key)
        return choices([b[0] for b in words_and_counts], weights=[b[1] for b in words_and_counts], k=1)[0]

    def random_value_from_random(self, keys):
        return self.random_value_from(self.random_key())

File: src/quix.ts
import {WorkerHttpvfs} from "sql.js-httpvfs"


interface KeysQueryResult {
  KEY: string;
}

interface ValueAndCountQueryResult {
  VALUE: string;
  COUNT: number;
}

interface ChapterQueryResult {
  CHAPTER: string;
}

export class QuixSql {
  worker: WorkerHttpvfs;
  _fragment: string = "";
  _currentKey?: string;
  _values: ValueAndCountQueryResult[] =[];
  _chapter?: string = "";
  _finalFragment: Boolean


  constructor(worker: WorkerHttpvfs) {
    this.worker = worker;
    this.randomKey()
      .then((key) => {
        this._fragment = key
        this._currentKey = key
      })
    this.values().then(() => console.log("constructor values are", this._values))
    this._finalFragment = false;

  }

    // def __init__(this, :):
    //     this.db = sqlite3.connect(db_path).cursor()
    //     this._fragment = this.randomKey()
    //     this._currentKey = this._fragment
    //     this._values = None
    //     this._chapter = None

    
   async generate_fragment() {
        var lc = 0
        debugger
        while ((!this._finalFragment) && ((await this.values()).length) > 0){
            lc += 1;
            console.log(lc)
            const randomIndex = Math.floor( Math.random() * this.values.length)
            const next_word = (await this.values()).splice(randomIndex, 1)[0].VALUE
            const possible_next_fragment = [this.fragment(), next_word].join(" ")
            if (await this.fragment_chapter(possible_next_fragment))
            {
                this._fragment = possible_next_fragment
                this._currentKey = undefined
                this._values = []
                this.generate_fragment()
            }
            else {
              continue
            }
          }
              
        this._finalFragment = true

        return {"fragment": this.fragment(), "chapter": (await this.fragment_chapter(this.fragment()) as string[])[0]}
      }

    fragment_chapter(fragment: string) {
        if (this._chapter) {
            return this.known_fragment_chapter(fragment)
        }
        return this.not_yet_known_fragment_chapter(fragment)
    }
    
    async not_yet_known_fragment_chapter(fragment: string) {
        const chapters =  await this.worker.db.query("SELECT chapter from chapters where text like ?", ["%"+fragment+"%"]) as string[]
        if (chapters.length === 1){
          this._chapter = chapters[0][0] 
        }
        return chapters ? chapters : undefined
      }
            

    async known_fragment_chapter(fragment: string) : Promise<string>  {
        const chapter = await this.worker.db.query("SELECT chapter from chapters where text like ? and chapter=?", [`%{fragment}%`, this._chapter]) as ChapterQueryResult[]
        return chapter[0].CHAPTER
    
    }

    async values() : Promise<ValueAndCountQueryResult[]> {
        if (this._values.length === 0) {
          this._values = await this.values_from(this.currentKey())
        }
        return this._values
      }

    currentKey() : string {
        if (this._currentKey === undefined) {
          const new_key_elements = this.fragment().split(' ')
          this._currentKey = new_key_elements.slice(new_key_elements.length -2).join(" ")
        }
          
        return this._currentKey
    }

    async randomKey() : Promise<string>{
        const key =  await this.worker.db.query("SELECT key from ENGLISH ORDER BY RANDOM() LIMIT 1") as KeysQueryResult[]
        return key[0].KEY
    }

    fragment(){
        return this._fragment
    }

    async values_from(key: string) : Promise<ValueAndCountQueryResult[]> {
        const values = await this.worker.db.query("SELECT value, count from ENGLISH where key=?", [key]) as ValueAndCountQueryResult[]
        return values
    }

    async random_value_from(key: string) : Promise<string> {
        const words_and_counts = await this.values_from(key)
        const weighted_words: string[] = words_and_counts.flatMap((value) => Array(value.COUNT).fill(value.VALUE))
        const randomIndex = Math.floor( Math.random() * weighted_words.length)
        return weighted_words[randomIndex]
    }

    async random_value_from_random() {
        return await this.random_value_from(await this.randomKey())
    }
  
}
